File: spicedb-java/examples/src/test/java/com/authzed/spicedb/examples/ReadRelationshipsTest.java
package com.authzed.spicedb.examples;

import com.authzed.spicedb.Filter;
import com.authzed.spicedb.Relationship;
import com.authzed.spicedb.Transaction;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.List;

import static com.authzed.spicedb.Consistency.*;
import static org.assertj.core.api.Assertions.*;

/**
 * Demonstrates reading relationships using {@link com.authzed.spicedb.SpiceDBClient#readRelationships}
 * with cursor-based auto-pagination.
 */
class ReadRelationshipsTest extends SpiceDBIntegrationTest {

    @BeforeEach
    void setUp() {
        client.deleteRelationships(Filter.of("document"));

        var txn = new Transaction();
        txn.touch(Relationship.of("document", "firstdoc", "viewer", "user", "alice"));
        txn.touch(Relationship.of("document", "firstdoc", "viewer", "user", "bob"));
        txn.touch(Relationship.of("document", "firstdoc", "editor", "user", "charlie"));
        client.write(txn);
    }

    @Test
    void reads_viewers_of_document() {
        Filter filter = Filter.of("document")
            .withResourceID("firstdoc")
            .withRelation("viewer");

        List<Relationship> relationships;
        try (var stream = client.readRelationships(full(), filter)) {
            relationships = stream.toList();
        }

        assertThat(relationships).hasSize(2);
        assertThat(relationships)
            .extracting(Relationship::subjectID)
            .containsExactlyInAnyOrder("alice", "bob");
    }

    @Test
    void reads_all_relations_on_document() {
        Filter filter = Filter.of("document")
            .withResourceID("firstdoc");

        List<Relationship> relationships;
        try (var stream = client.readRelationships(full(), filter)) {
            relationships = stream.toList();
        }

        assertThat(relationships).hasSize(3);
        assertThat(relationships)
            .extracting(Relationship::resourceRelation)
            .contains("viewer", "editor");
    }

    @Test
    void empty_result_for_nonexistent_resource() {
        Filter filter = Filter.of("document")
            .withResourceID("nonexistent");

        try (var stream = client.readRelationships(full(), filter)) {
            assertThat(stream.count()).isZero();
        }
    }
}

File: spicedb-csharp/examples/ReadRelationships/ReadRelationshipsTest.cs
using Xunit;
// Example ReadRelationships demonstrates reading relationships with an
// async enumerable.

using SpiceDB.Client;
using static SpiceDB.Client.Consistency;

namespace ReadRelationships;

public class ReadRelationshipsTest
{
    private const string Schema = """
        definition user {}

        definition document {
            relation viewer: user
            relation editor: user
            permission view = viewer + editor
            permission edit = editor
        }
        """;

    [Fact]
    public async Task ReadRelationships_ReturnsMatchingRelationships()
    {
        await using var client = SpiceDBClient.CreatePlaintext(SpiceDBTestServer.Endpoint, SpiceDBTestServer.Token);

        await client.WriteSchemaAsync(Schema);

        var txn = new Transaction();
        txn.Touch(Relationship.FromTriple("document", "firstdoc", "viewer", "user", "alice"));
        txn.Touch(Relationship.FromTriple("document", "firstdoc", "viewer", "user", "bob"));
        await client.WriteAsync(txn);

        // Read relationships
        var filter = new Filter("document")
            .WithResourceID("firstdoc")
            .WithRelation("viewer");

        var relationships = new List<Relationship>();
        await foreach (var rel in client.ReadRelationshipsAsync(Full(), filter))
        {
            relationships.Add(rel);
        }

        Assert.True(relationships.Count >= 2, "expected at least two viewer relationships");

        var subjectIDs = relationships.Select(r => r.SubjectID).ToHashSet();
        Assert.Contains("alice", subjectIDs);
        Assert.Contains("bob", subjectIDs);
    }
}
